---
title: "Use case 1: Region-based association study"
output: html_notebook
---

First, we load the library:

```{r}
# Import CASMAP package
library(sigpatsearch)
```

Next, we create an object that will search using the `FACS` method:

```{r}
# Create object to search for high-order epistasis in a Genome-Wide Association Study (GWAS)
high_order_epistasis <- sigpatsearch(method='FACS')

# Set hyperparameters
high_order_epistasis$set_alpha(0.05)  # Target FWER
high_order_epistasis$set_lmax(0)  # Include interactions of any order in the analysis
```

First, we get the paths to the data files:
```{r}

# Path to input files (genotype, phenotype and categorical covariate)
dataset = 'avrB'


# Get current working directory. Used to set the output directory relative to the current one.
current_dir = getwd()

#now create paths
data_path <- file.path(current_dir, '..', 'data', 'high_order_epistasis', dataset)

#and get file names
genotype_file  <- file.path(data_path, 'X.dat')
phenotype_file <- file.path(data_path, 'Y.dat')
covariate_file <- file.path(data_path, 'C.dat')
```

Next we read the data:
```{r}
# Read input files
high_order_epistasis$read_eth_files(genotype_file, phenotype_file, covariate_file)
```

Then the next step runs the data:
```{r, eval=FALSE}
# Run significant pattern mining algorithm to retrieve statistically associated interactions between genomic variants
high_order_epistasis$execute()
```


Then the next step runs the data:
```{r, eval=FALSE}
# Run significant pattern mining algorithm to retrieve statistically associated interactions between genomic variants
high_order_epistasis$execute()
```

Save the results to files:
```{r, eval=FALSE}
# Path to output directory
# Get current working directory. Used to set the output directory relative to the current one.
current_dir = getwd()
#obtain the path separator, since Windows uses "\" and Unix-based systems use "/"
os_sep <- .Platform$file.sep
#output_path <- (current_dir, '..', os_sep, 'output', os_sep, 'high_order_epistasis', os_sep, dataset, os_sep)
output_path <- file.path(current_dir, '..', 'output', 'high_order_epistasis', dataset)

# Write high-level summary and profiling info related to the execution of the algorithm
summary_path <- file.path(output_path, 'summary.txt')
high_order_epistasis.write_summary(summary_path)

profiling_path <- file.path(output_path, 'profiling.txt')
high_order_epistasis.write_profile(profiling_path)
```


